implementations/rust/message/src/lib.rs

```diff
@@ -1,517 +1,593 @@
+                (HostAddressType::Ipv4, addr) => {
+                    let ip4 = Ipv4Addr::new(addr[0], addr[1], addr[2], addr[3]);
+                    let ip_addr = IpAddr::V4(ip4);
+                    Ok((ip_addr, &u[5..]))
+                }
+                _ => Err("".to_string()),
             }
-            _ => Err("".to_string()),
         }
     }
-}
 
-impl Codec for SocketAddr {
-    type Inner = SocketAddr;
-    fn encode(sock: SocketAddr, v: &mut Vec<u8>) -> Result<(), String> {
-        match sock {
-            std::net::SocketAddr::V4(sock4) => {
-                v.push(HostAddressType::Ipv4 as u8);
-                v.extend_from_slice(sock4.ip().octets().as_ref());
-                let p = sock4.port();
-                v.extend_from_slice(&p.to_le_bytes());
+    impl Codec for SocketAddr {
+        type Inner = SocketAddr;
+        fn encode(sock: SocketAddr, v: &mut Vec<u8>) -> Result<(), String> {
+            match sock {
+                std::net::SocketAddr::V4(sock4) => {
+                    v.push(HostAddressType::Ipv4 as u8);
+                    v.extend_from_slice(sock4.ip().octets().as_ref());
+                    let p = sock4.port();
+                    v.extend_from_slice(&p.to_le_bytes());
+                }
+                std::net::SocketAddr::V6(sock6) => {
+                    v.push(HostAddressType::Ipv6 as u8);
+                    v.extend_from_slice(sock6.ip().octets().as_ref());
+                    let p = sock6.port();
+                    v.extend_from_slice(&p.to_le_bytes());
+                }
             }
-            std::net::SocketAddr::V6(sock6) => {
-                v.push(HostAddressType::Ipv6 as u8);
-                v.extend_from_slice(sock6.ip().octets().as_ref());
-                let p = sock6.port();
-                v.extend_from_slice(&p.to_le_bytes());
+            Ok(())
+        }
+        fn decode(u: &[u8]) -> Result<(SocketAddr, &[u8]), String> {
+            match (HostAddressType::try_from(u[0])?, &u[1..]) {
+                (HostAddressType::Ipv4, addr) => {
+                    let ip4 = Ipv4Addr::new(addr[0], addr[1], addr[2], addr[3]);
+                    let port = u16::from_le_bytes([addr[4], addr[5]]);
+                    let sock = SocketAddr::new(IpAddr::V4(ip4), port);
+                    Ok((sock, &addr[6..]))
+                }
+                _ => Err("".to_string()),
             }
         }
-        Ok(())
     }
-    fn decode(u: &[u8]) -> Result<(SocketAddr, &[u8]), String> {
-        match (HostAddressType::try_from(u[0])?, &u[1..]) {
-            (HostAddressType::Ipv4, addr) => {
-                let ip4 = Ipv4Addr::new(addr[0], addr[1], addr[2], addr[3]);
-                let port = u16::from_le_bytes([addr[4], addr[5]]);
-                let sock = SocketAddr::new(IpAddr::V4(ip4), port);
-                Ok((sock, &addr[6..]))
+
+    impl Route {
+        pub fn print_route(&self) {
```

Comment:
May benefit from implementing `std::fmt::Display` for the different route address types, so anyone can `println!` them.